### ISSUE TITLE
Regression in swept disk solid in v0.8

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#5473]
0.7 looks like this:

![2024-09-24-224558_547x267_scrot](https://github.com/user-attachments/assets/812d8008-18cc-4c2d-af47-752ecc1ca22e)

0.8 looks like this:

![2024-09-24-224607_825x421_scrot](https://github.com/user-attachments/assets/0b86aba6-4881-49db-a6bb-f4697952ae9c)

```
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition [CoordinationView]'),'2;1');
FILE_NAME('','2024-09-24T22:38:00',(''),(''),'IfcOpenShell 0.8.1-a51ed6e6b','IfcOpenShell 0.8.1-a51ed6e6b','');
FILE_SCHEMA(('IFC4'));
ENDSEC;
DATA;
#1=IFCPROJECT('3ZHfd54197sOvwWETRXIhF',$,'project_name',$,$,$,$,(#11),#6);
#2=IFCDIMENSIONALEXPONENTS(1,0,0,0,0,0,0);
#3=IFCSIUNIT(*,.LENGTHUNIT.,$,.METRE.);
#4=IFCMEASUREWITHUNIT(IFCREAL(0.3048),#3);
#5=IFCCONVERSIONBASEDUNIT(#2,.LENGTHUNIT.,'foot',#4);
#6=IFCUNITASSIGNMENT((#3));
#7=IFCCARTESIANPOINT((0.,0.,0.));
#8=IFCDIRECTION((0.,0.,1.));
#9=IFCDIRECTION((1.,0.,0.));
#10=IFCAXIS2PLACEMENT3D(#7,#8,#9);
#11=IFCGEOMETRICREPRESENTATIONCONTEXT($,'Model',3,1.E-05,#10,$);
#12=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Body','Model',*,*,*,*,#11,$,.MODEL_VIEW.,$);
#13=IFCSITE('3TC2wPUMv5P9HYlVOMiPUK',$,'site_name',$,$,$,$,$,$,$,$,$,$,$);
#14=IFCBUILDING('2guf2ZNR94FQSt29s79yzN',$,'building_name',$,$,$,$,$,$,$,$,$);
#15=IFCBUILDINGSTOREY('2j$XMQ58v9sANy9r5v7oI$',$,'storey_name',$,$,$,$,$,$,$);
#16=IFCRELAGGREGATES('2jNFcJkbv2K99zO3j5$gQh',$,$,$,#1,(#13));
#17=IFCRELAGGREGATES('3GOzceF$bCqwTwgyib_x$C',$,$,$,#13,(#14));
#18=IFCRELAGGREGATES('1h84f8J8HCHgceh4Ul$O8B',$,$,$,#14,(#15));
#19=IFCPIPESEGMENT('3I4NFpEU54jPPZef$KxnxS',$,$,$,$,#35,#29,$,$);
#25=IFCCARTESIANPOINTLIST3D(((0.,0.,0.),(10.,0.,0.),(17.1,2.9,0.),(20.,10.,0.),(20.,20.,0.)));
#26=IFCINDEXEDPOLYCURVE(#25,(IFCLINEINDEX((1,2)),IFCARCINDEX((2,3,4)),IFCLINEINDEX((4,5))),$);
#27=IFCSWEPTDISKSOLID(#26,0.5,$,$,$);
#28=IFCSHAPEREPRESENTATION(#12,'Body','AdvancedSweptSolid',(#27));
#29=IFCPRODUCTDEFINITIONSHAPE($,$,(#28));
#30=IFCRELCONTAINEDINSPATIALSTRUCTURE('2xcgOq4gjC9hyfdAJa3byb',$,$,$,(#19),#15);
#31=IFCCARTESIANPOINT((0.,0.,0.));
#32=IFCDIRECTION((0.,0.,1.));
#33=IFCDIRECTION((1.,0.,0.));
#34=IFCAXIS2PLACEMENT3D(#31,#32,#33);
#35=IFCLOCALPLACEMENT($,#34);
ENDSEC;
END-ISO-10303-21;
```

**Comment by aothms** [comment]
In v0.7 we had a ton of workarounds for sweeps along curves, which have always been rather buggy in occt. A lot of the files in https://github.com/ifcopenshell/files are actually issues pertaining sweeps along curves. But these workarounds also caused artefacts (mostly witnessed by @brunopostle who would use these for roof trims over various discontinuous wires). So I was hoping that with the OCCT version upgrades that have been going on, these workarounds were no longer necessary, or at least  to bring them back only on a case by case basis. This is one of these cases, that luckily can be selectively enabled and is without artefacts. Looks really nasty though this one, no idea how something so obvious can be so obviously broken in occt, maybe I'm holding it wrong.

**Comment by Moult** [comment]
@aothms given the popularity of swept disk solid and prevalence esp. in rebar, as well as how slow (and apparently buggy, as you say) OCC is with sweeps, would it make sense to implement swept disk in CGAL? Or even just as a manual "microkernel" as you've hinted once or twice - where I can think of a very naive faceted approach to sweeping a bunch of verts.

**Comment by aothms** [comment]
You're right. It's really not that hard in the simple ifc case, the complexity in OCCT most likely comes from the abstractness of their implementation, i.e sweeping arbitrary faces over arbitrary curves restricted to arbitrary surfaces (i.e. IfcSurfaceCurveSweptAreaSolid).

We also kind of already have a precedent for this in how we solve the infrastructure sweeps.

But we probably first need to have a much more intelligent and configurable hybrid kernel for that and/or the ability to comvert kernel items from one geomlib to the other.

**Comment by BIMvoice** [comment]
This looks resolved on the current v0.8 line. The swept-disk path now selects the transition mode based on the directrix: `sweep_along_curve.cpp` uses BRepBuilderAPI_Transformed for a C1-continuous directrix containing circular arcs (this sample) and RightCorner otherwise, which produces a clean pipe for the line+arc+line case.

Verified with a local build (OpenCASCADE 7.9.2) on your inline sample: the swept disk renders as a watertight manifold solid with volume 28.04, matching the expected pi*r*r*L (r=0.5, directrix length ~35.75) of ~28.08. Recommend closing as fixed unless anyone still sees the distortion on an older OpenCASCADE.
